File: Grind75/spiralOrder.py
from typing import List

def spiralOrder(matrix: List[List[int]]) -> List[int]:
    result: List[int] = []
    if not matrix or not matrix[0]:
        return result

    m, n = len(matrix), len(matrix[0])
    top, bottom = 0, m - 1
    left, right = 0, n - 1

    # Continue until all boundaries have been crossed
    while top <= bottom and left <= right:
        # 1. Traverse from left to right across the top row
        for col in range(left, right + 1):
            result.append(matrix[top][col])
        top += 1

        # 2. Traverse from top to bottom down the right column
        for row in range(top, bottom + 1):
            result.append(matrix[row][right])
        right -= 1

        # 3. If there's still a bottom row remaining, traverse from right to left
        if top <= bottom:
            for col in range(right, left - 1, -1):
                result.append(matrix[bottom][col])
            bottom -= 1

        # 4. If there's still a left column remaining, traverse from bottom to top
        if left <= right:
            for row in range(bottom, top - 1, -1):
                result.append(matrix[row][left])
            left += 1

    return result


# Time Complexity: O(m * n), since each element is visited exactly once.
# Space Complexity: O(1) extra space (excluding the output list).

File: Grind75/spiralOrder.js
function spiralOrder(matrix) {
  const result = [];
  if (matrix.length === 0) return result;

  let top = 0;
  let bottom = matrix.length - 1;
  let left = 0;
  let right = matrix[0].length - 1;

  while (top <= bottom && left <= right) {
    // traverse top row left to right
    for (let j = left; j <= right; j++) {
      result.push(matrix[top][j]);
    }
    top++;

    // traverse right column top to bottom
    for (let i = top; i <= bottom; i++) {
      result.push(matrix[i][right]);
    }
    right--;

    // traverse bottom row right to left (if still valid)
    if (top <= bottom) {
      for (let j = right; j >= left; j--) {
        result.push(matrix[bottom][j]);
      }
      bottom--;
    }

    // traverse left column bottom to top (if still valid)
    if (left <= right) {
      for (let i = bottom; i >= top; i--) {
        result.push(matrix[i][left]);
      }
      left++;
    }
  }

  return result;
}

// Time complexity: O(m * n) - every element is visited exactly once
// Space complexity: O(1) - excluding the output array, only constant extra space used for boundary pointers
